fullstory-segment-middleware/src/test/java/com/fullstorydev/fullstory_segment_middleware/FullStorySegmentMiddlewareTest.java

```diff
@@ -0,0 +1,442 @@
+    }
+
+    @Test
+    public void intercept_TrackPayloadChain_AllowlistAllTrackEvents_ChainProceedCalled() {
+        fullStorySegmentMiddleware = new FullStorySegmentMiddleware(mockContext, "SegmentWriteMockKey");
+        fullStorySegmentMiddleware.allowlistAllTrackEvents = true;
+        fullStorySegmentMiddleware.enableFSSessionURLInEvents = false;
+
+        TrackPayload trackPayload = new TrackPayload.Builder().userId("userId").event("event").build();
+        when(mockChain.payload()).thenReturn(trackPayload);
+        fullStorySegmentMiddleware.intercept(mockChain);
+
+        // chain.proceed is called with the correct payload
+        verify(mockChain, times(1)).proceed(trackPayload);
+    }
+
+    @Test
+    public void intercept_AliasPayload_DefaultCase() {
+        fullStorySegmentMiddleware = new FullStorySegmentMiddleware(mockContext, "SegmentWriteMockKey");
+        fullStorySegmentMiddleware.allowlistAllTrackEvents = true;
+        fullStorySegmentMiddleware.enableFSSessionURLInEvents = false;
+
+        AliasPayload aliasPayload = new AliasPayload.Builder().previousId("previousId").userId("userId").build();
+        when(mockChain.payload()).thenReturn(aliasPayload);
+        fullStorySegmentMiddleware.intercept(mockChain);
+
+        // chain.proceed is called with the correct payload
+        verify(mockChain, times(1)).proceed(aliasPayload);
+    }
+
+    @Test
+    public void intercept_EnableFSSessionURLInEvents_VerifyChainProceed() {
+        fullStorySegmentMiddleware = new FullStorySegmentMiddleware(mockContext, "SegmentWriteMockKey");
+        fullStorySegmentMiddleware.enableFSSessionURLInEvents = true;
+
+        TrackPayload inputTrackPayload = new TrackPayload.Builder().userId("userId").event("event").build();
+        when(mockChain.payload()).thenReturn(inputTrackPayload);
+        fullStorySegmentMiddleware.intercept(mockChain);
+
+        // chain.proceed is called with the correct payload
+        Map<String, Object> newContext = new HashMap<>(inputTrackPayload.context());
+        newContext.put("fullstoryUrl", FS.getCurrentSessionURL());
+        Map<String, Object> newProperties = new HashMap<>(inputTrackPayload.properties());
+        newProperties.put("fullstoryUrl", FS.getCurrentSessionURL());
+
+        TrackPayload newTrackPayload = inputTrackPayload.toBuilder()
+                .context(newContext)
+                .properties(newProperties)
+                .build();
+
+        verify(mockChain, times(1)).proceed(newTrackPayload);
+    }
+
+    @Test
+    public void addFSUrlToContext_AddsURLToContext() {
+        fullStorySegmentMiddleware = new FullStorySegmentMiddleware(mockContext, "SegmentWriteMockKey");
+        Map<String, Object> input = new HashMap<>();
+        fullStorySegmentMiddleware.addFSUrlToContext(input);
+
+        Map<String, Object> expect = new HashMap<>();
+        expect.put("fullstoryUrl", FS.getCurrentSessionURL());
```

Comment:
this works! Future research could be to see if you can _inject_ FS somehow into the middleware using an Inversion of Control/Dependency Injection pattern. Then you can mock the FS methods and verify that they were called, rather than compare values. Might not be possible given how the middleware and FS are architected.